// @uiw/react-textarea-code-editor/lib/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {};
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _utils = require("./utils");

var _shortcuts = _interopRequireDefault(require("./shortcuts"));

var styles = _interopRequireWildcard(require("./styles"));

var _SelectionText = require("./SelectionText");

Object.keys(_SelectionText).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _SelectionText[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _SelectionText[key];
    }
  });
});
var _excluded = ["prefixCls", "value", "padding", "minHeight", "placeholder", "language", "className", "style", "onChange"];

var _default = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _props$prefixCls = props.prefixCls,
      prefixCls = _props$prefixCls === void 0 ? 'w-tc-editor' : _props$prefixCls,
      _ = props.value,
      _props$padding = props.padding,
      padding = _props$padding === void 0 ? 10 : _props$padding,
      _props$minHeight = props.minHeight,
      minHeight = _props$minHeight === void 0 ? 16 : _props$minHeight,
      placeholder = props.placeholder,
      language = props.language,
      className = props.className,
      style = props.style,
      _onChange = props.onChange,
      other = (0, _objectWithoutProperties2.default)(props, _excluded);

  var _useState = (0, _react.useState)(props.value || ''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  (0, _react.useEffect)(function () {
    return setValue(props.value || '');
  }, [props.value]);
  var textRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, function () {
    return textRef.current;
  });
  var contentStyle = {
    paddingTop: padding,
    paddingRight: padding,
    paddingBottom: padding,
    paddingLeft: padding
  };
  var htmlStr = (0, _react.useMemo)(function () {
    return (0, _utils.processHtml)("<pre aria-hidden=true><code ".concat(language ? "class=\"language-".concat(language, "\"") : '', " >").concat((0, _utils.htmlEncode)(String(value || '')), "</code><br /></pre>"));
  }, [value, language]);
  var preView = (0, _react.useMemo)(function () {
    return /*#__PURE__*/_react.default.createElement("div", {
      style: (0, _objectSpread2.default)((0, _objectSpread2.default)((0, _objectSpread2.default)({}, styles.editor), contentStyle), {}, {
        minHeight: minHeight
      }),
      className: "".concat(prefixCls, "-preview ").concat(language ? "language-".concat(language) : ''),
      dangerouslySetInnerHTML: {
        __html: htmlStr
      }
    });
  }, // eslint-disable-next-line react-hooks/exhaustive-deps
  [prefixCls, language, htmlStr]);
  return /*#__PURE__*/_react.default.createElement("div", {
    style: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, styles.container), style),
    className: "".concat(prefixCls, " ").concat(className || '')
  }, /*#__PURE__*/_react.default.createElement("textarea", (0, _extends2.default)({
    autoComplete: "off",
    autoCorrect: "off",
    spellCheck: "false",
    autoCapitalize: "off"
  }, other, {
    placeholder: placeholder,
    onKeyDown: function onKeyDown(evn) {
      (0, _shortcuts.default)(evn);
      other.onKeyDown && other.onKeyDown(evn);
    },
    style: (0, _objectSpread2.default)((0, _objectSpread2.default)((0, _objectSpread2.default)((0, _objectSpread2.default)({}, styles.editor), styles.textarea), contentStyle), {}, {
      minHeight: minHeight
    }, placeholder && !value ? {
      WebkitTextFillColor: 'inherit'
    } : {}),
    ref: textRef,
    onChange: function onChange(evn) {
      setValue(evn.target.value);
      _onChange && _onChange(evn);
    },
    className: "".concat(prefixCls, "-text"),
    value: value
  })), preView);
});

exports.default = _default;
